src/librustc/middle/mem_categorization.rs

```diff
@@ -703,7 +703,13 @@ impl<'a, 'gcx, 'tcx> MemCategorizationContext<'a, 'gcx, 'tcx> {
 
         let kind = match self.tables.closure_kinds.get(&fn_node_id) {
             Some(&(kind, _)) => kind,
-            None => span_bug!(span, "missing closure kind")
+            None => {
+                let ty = self.node_ty(fn_node_id)?;
+                match ty.sty {
+                    ty::TyGenerator(..) => ty::ClosureKind::FnOnce,
```

Comment:
The environment of generators are taken by value until after MIR drop elaboration. I don't think MC can model `&move` and I don't think it matters for borrow checking.

I'm not seeing the reason we need to model the deref of the closure environment at all, so I might be missing something.